### PR TITLE
Docker BoM

### PR DESCRIPTION
**Pull request by qrdl**
- Support for dependencies inside docker images (with `docker://` scheme)
- CycloneDX JSON and XML output formats for BoM (https://cyclonedx.org/)
- package URLs (https://github.com/package-url/purl-spec)

### CODE REVIEW

File: docs/cmd/vcn_untrust.md [hunk 1]
```diff
@@ -52,29 +52,24 @@ echo my-file | vcn n -
...
-      --no-ignore-file        if set, .vcnignore will be not written inside the targeted dir (affects dir:// only)
-  -p, --public                when notarized as public, the asset name and metadata will be visible to everyone
-      --read-only             if set, no files will be written into the targeted dir (affects dir:// only)
-  -r, --recursive             if set, wildcard usage will walk inside subdirectories of provided path
+      --attach vcn a binary1 --attach=vscanner.result:jobid123   add user defined file attachments. Ex. vcn n myfile --attach mysecondfile. (repeat --attach for multiple entries). It's possible to specify a label for each entry, Ex: --attach=vscanner.result:jobid123. In this way it will be possible to retrieve the specific attachment with vcn a binary1 --attach=vscanner.result:jobid123 or `vcn a binary1 --attach=jobid123` to get all attachments
+  -a, --attr key=value                                           add user defined attributes (repeat --attr for multiple entries)
+      --bom-cascade                                              cascade the operation to all assets that include the asset being processed
```
Comment (mmeloni) [inline docs/cmd/vcn_untrust.md:57]: It's possible to add an example?

Comment (qrdl) [reply docs/cmd/vcn_untrust.md:57]: https://github.com/codenotary/vcn/blob/feat/bom/docs/user-guide/bom.md#cascade-operations

File: pkg/bom/docker/docker.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+				projected := int(fileCount * float64(len(a.binaries)) / binCount)
+				bar.ChangeMax(projected)
+			}
+
+			scanner := bufio.NewScanner(bytes.NewBuffer(stdOut))
+			for scanner.Scan() {
+				fields := strings.Split(scanner.Text(), " ")
+				var lib string
+				switch len(fields) {
+				case 2:
```
Comment (mmeloni) [inline pkg/bom/docker/docker.go:121]: missing default?

File: pkg/bom/docker/docker.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+					lib = strings.TrimSpace(fields[2])
+				}
+
+				if bar != nil {
+					bar.Add(1)
+				}
+				dep, err := a.pkg.PackageByFile(a.ex, lib, output)
+				if err != nil {
+					// it is possible that lib is a symlink, in this case package manager can't find it, so resolve it and retry
+					stdOut, stdErr, exitCode, err = a.ex.Exec([]string{"readlink", "-f", lib})
```
Comment (mmeloni) [inline pkg/bom/docker/docker.go:133]: This is done for each error? It's possible to detect only the symlink case?

Comment (qrdl) [reply pkg/bom/docker/docker.go:133]: Good idea. I'll change it

File: pkg/bom/docker/docker.go (hunk 2 above)
Comment (qrdl) [reply pkg/bom/docker/docker.go:122]: yep. fixed

File: pkg/bom/docker/docker.go (hunk 3 above)
Comment (qrdl) [reply pkg/bom/docker/docker.go:137]: fixed

**Review by mmeloni — APPROVED** [review]

**Pull request merged**